src/core/utils/git.ts

```diff
@@ -0,0 +1,75 @@
+import { execSync } from "node:child_process";
+import path from "node:path";
+import type { Branch, Project } from "../db/schema";
+import { getErrorMessage } from "./error";
+
+export class GitError extends Error {
+  constructor(message: string) {
+    super(message);
+    this.name = "GitError";
+  }
+}
+
+export function execGit(command: string): string {
+  try {
+    return execSync(`git ${command}`, { encoding: "utf-8" }).trim();
+  } catch (error) {
+    throw new GitError(`Git command failed: ${getErrorMessage(error)}`);
+  }
+}
+
+export function getProjectMetadata(): Project {
+  try {
+    const repoPath = execGit("rev-parse --show-toplevel");
+    const repoName = path.basename(repoPath);
+
+    return {
+      id: repoPath, // TODO: make it unique
+      repoPath,
+      repoName,
+    };
+  } catch (error) {
+    throw new GitError(
+      `Failed to get project metadata: ${getErrorMessage(error)}`,
+    );
+  }
+}
+
+export function getBranchMetadata(projectId: string): Branch {
+  try {
+    const branchName = execGit("rev-parse --abbrev-ref HEAD");
+
+    if (branchName === "HEAD") {
+      throw new GitError("Cannot determine branch name in detached HEAD state");
+    }
+
+    const branchRefHashId = execGit(
+      `rev-parse --symbolic-full-name ${branchName}`,
+    );
+
```

Comment:
https://github.com/offlegacy/git-intent/pull/30#discussion_r2187897800

It would also be good to discuss how to handle cases where the branch name changes. @jcha0713

Reply:
How about prompting the user when a branch with stored intents isn't found? We could probably show branches with the most recent activity using `git reflog` and let them choose or ask them to clear up the intents.